core/src/main/java/org/elasticsearch/index/query/RangeQueryBuilder.java

```diff
@@ -259,37 +261,54 @@ public String getWriteableName() {
-            switch (relation) {
-            case DISJOINT:
-                return new MatchNoneQueryBuilder();
-            case WITHIN:
-                FieldStats<?> fieldStats = fieldStatsProvider.get(fieldName);
-                if (!(fieldStats.getMinValue().equals(from) && fieldStats.getMaxValue().equals(to) && includeUpper && includeLower)) {
-                    // Rebuild the range query with the bounds for this shard.
-                    // The includeLower/Upper values are preserved only if the
-                    // bound has not been changed by the rewrite
-                    RangeQueryBuilder newRangeQuery = new RangeQueryBuilder(fieldName);
-                    String dateFormatString = format == null ? null : format.format();
-                    newRangeQuery.from(fieldStats.getMinValue(), includeLower || fieldStats.match(
-                            new IndexConstraint(fieldName, Property.MIN, Comparison.GT, fieldStats.stringValueOf(from, dateFormatString))));
-                    newRangeQuery.to(fieldStats.getMaxValue(), includeUpper || fieldStats.match(
-                            new IndexConstraint(fieldName, Property.MAX, Comparison.LT, fieldStats.stringValueOf(to, dateFormatString))));
-                    newRangeQuery.format = format;
-                    newRangeQuery.timeZone = timeZone;
-                    return newRangeQuery;
-                } else {
-                    return this;
+    protected QueryBuilder<?> doRewrite(QueryRewriteContext queryRewriteContext) throws IOException {
+        MapperService mapperService = queryRewriteContext.getMapperService();
+        if (mapperService != null) {
+            FieldStatsProvider fieldStatsProvider = queryRewriteContext.getFieldStatsProvider();
+            // If the fieldStatsProvider is null we are not on the shard and cannot
+            // rewrite so just return without rewriting
+            if (fieldStatsProvider != null) {
+                DateMathParser dateMathParser = format == null ? null : new DateMathParser(format);
+                String dateFormatString = format == null ? null : format.format();
+                Object from = this.from;
+                Object to = this.to;
+                MappedFieldType fieldType = mapperService.fullName(fieldName);
+                if (fieldType instanceof DateFieldType) {
+                    // If the field is a date field resolve the date math now to
+                    // prevent having to do it multiple times later
+                    DateFieldType dateFieldType = (DateFieldType) fieldType;
+                    if (from instanceof BytesRef) {
+                        from = dateFieldType.parseToMilliseconds(from, !includeLower, timeZone, dateMathParser);
+                    }
+                    if (to instanceof BytesRef) {
+                        to = dateFieldType.parseToMilliseconds(to, includeUpper, timeZone, dateMathParser);
+                    }
+                }
```

Comment:
It could be but the new values of from and to are also used below in the WITHIN case to work out what to do with the includeLower and includeUpper values so we would have to do it there anyway